SPEC.md

```diff
@@ -0,0 +1,258 @@
+## Apollo Fetchers
+
+The purpose of apollo-fetcher is to define an extensible standard interface for fetching GraphQL results. To that end, apollo-fetcher describes an interface containing a single method that connects a GraphQL `Operation` (also called request), to an `Observable` which delivers the results of the operation. Operations that return a single result can easily be mapped to a promise, since the observable's `next` function is called only once. In the general case, Apollo Fetcher uses observables to support GraphQL subscriptions and live queries. The apollo-fetcher's `Observable` follows the ECMAScript [proposal](https://github.com/tc39/proposal-observable). A basic fetcher is visualized as follows:
+
+<p align="center">
+  <br>
+  <img src="images/apollo-fetcher.png" alt="Apollo Fetcher"/>
+</p>
+
+Apollo Fetchers are designed to be used as a stand-alone extensible GraphQL client. In addition, fetchers fit all current and anticipated needs of Apollo Client. As such, fetchers are targetted towards the following use-cases:
+
+* Do a simple GraphQL query with it that resolves only once
+* Do a simple GraphQL mutation that resolves only once
+* Do a subscription
+* Do a GraphQL query that contains @defer, @stream or @live
+* Send extra metadata along with queries
+* Receive extra metadata with the query result
+* Support middleware to modify headers
+* Support middleware to modify query, variables, operationName and context
+* Isomorphic: It works in the browser and on the server
+* Support afterware to modify response
+* Support HTTP transport, websocket transport, hybrid transport
+* Support alternative transports
+* Support per request retries based on network status or response
+* Support polling
+* Support batching
+* Request deduplication of on-the-wire requests
+* Support splitting: eg. send queries and mutations to different servers
+
+If you can think of more use-cases, please open a PR adding to this list and make sure to clearly explain your use-case.
+
+In order to support all these use cases with a simple fetcher design, we want to use a modular architecture that composition of fetchers to stack functionality. An Intermediate Fetcher that modifies the operation passed down or the response received is represented below:
+
+<p align="center">
+  <br>
+  <img src="images/generic-stack.png" alt="Generic Fetcher Stack"/>
+</p>
+
+
+## Examples
+
+http-fetcher sends an operation over HTTP to a server at a constructor specified URI. Additionally, the constructor accepts an optional fetch function and defaults to isomorphic-fetch. A custom fetch function would contain the HTTP specific behavior normally found in middleware or afterware. In this manner, http-fetcher can function as a stand-alone GraphQL client:
+
+<p align="center">
+  <br>
+  <img src="images/http-fetcher.png" alt="Http Fetcher"/>
+</p>
+
+```js
+const httpFetcher = new HTTPFetcher({ uri: 'http://api.githunt.com/graphql' });
```

Comment:
This is a personal preference, but I prefer to avoid using `new` where possible. It can be confusing to beginners to know when to use it and aside from `classes`, is not needed for standard function calls.